sqlmesh/core/engine_adapter/spark.py

```diff
@@ -50,37 +51,137 @@ def _use_spark_session(self) -> bool:
-            exp.DataType.Type.TEXT: spark_types.StringType,
+        PRIMITIVE_MAPPING: t.Dict[t.Any, t.Any] = {
+            exp.DataType.Type.TINYINT: spark_types.ByteType,
+            exp.DataType.Type.SMALLINT: spark_types.ShortType,
             exp.DataType.Type.INT: spark_types.IntegerType,
             exp.DataType.Type.BIGINT: spark_types.LongType,
             exp.DataType.Type.FLOAT: spark_types.FloatType,
             exp.DataType.Type.DOUBLE: spark_types.DoubleType,
             exp.DataType.Type.DECIMAL: spark_types.DecimalType,
+            exp.DataType.Type.TEXT: spark_types.StringType,
+            # SQLGlot currently converts these to Strings
+            # exp.DataType.Type.VARCHAR: spark_types.VarcharType,
+            # exp.DataType.Type.CHAR: spark_types.CharType,
+            exp.DataType.Type.BINARY: spark_types.BinaryType,
+            exp.DataType.Type.BOOLEAN: spark_types.BooleanType,
             exp.DataType.Type.DATE: spark_types.DateType,
+            exp.DataType.Type.DATETIME: spark_types.TimestampNTZType,
+            exp.DataType.Type.TIMESTAMPLTZ: spark_types.TimestampType,
+            exp.DataType.Type.TIMESTAMPTZ: spark_types.TimestampType,
             exp.DataType.Type.TIMESTAMP: spark_types.TimestampType,
         }
 
-        try:
-            return spark_types.StructType(
-                [
-                    spark_types.StructField(col_name, mapping[col_type.this]())
-                    for col_name, col_type in columns_to_types.items()
+        COMPLEX_TYPE_MAPPING: t.Dict[t.Any, t.Any] = {
+            exp.DataType.Type.ARRAY: spark_types.ArrayType,
+            exp.DataType.Type.MAP: spark_types.MapType,
+            exp.DataType.Type.STRUCT: spark_types.StructType,
+        }
+
+        SPARK_PRIMITIVE_MAPPING = {v: k for k, v in PRIMITIVE_MAPPING.items()}
+        SPARK_COMPLEX_TYPE_MAPPING = {v: k for k, v in COMPLEX_TYPE_MAPPING.items()}
+
+        def sqlglot_complex_to_spark_complex(complex_type: exp.DataType) -> spark_types.DataType:
+            is_struct = complex_type.is_type(exp.DataType.Type.STRUCT)
+            expressions = []
+            for column_def in complex_type.expressions:
+                col_name = column_def.this.name if is_struct else None
+                data_type = column_def.args["kind"] if is_struct else column_def
+                primitive_func = PRIMITIVE_MAPPING.get(data_type.this)
+                type_func = (
+                    primitive_func
+                    if primitive_func
+                    else partial(sqlglot_complex_to_spark_complex, data_type)
+                )
+                if is_struct:
+                    expressions.append(spark_types.StructField(col_name, type_func()))
+                else:
+                    expressions.append(type_func())
+            klass = COMPLEX_TYPE_MAPPING[complex_type.this]
+            return klass(*([expressions] if is_struct else expressions))
```

Comment:
what's going on here? why do you have a nested array? can you add kwargs? klass(expressions=) ?

Reply:
This is not a SQLGlot class it is spark. Array/Map want args to be passed in while struct wants a single list to be passed in. I rewrote it a bit if that helps.